### PR TITLE
[ENH] Add `check_causal_discovery` method to test compliance of causal algorithms (#2839)

### LINKED ISSUE #2839: [ENH] Add method to test whether causal discovery method is complaint
We have a unified interface for all the causal discovery algorithms (such as they must have fit and score methods; `fit` method must set the `causal_graph_` attribute). The causal discovery extension template `devtools/extension_templates/_causal_discovery.py` outlines all of these requirements.

To make it easier to check whether a new causal discovery algorithm is compliant with our interface, we should write a test method that takes in a causal discovery class instance and runs all the interface checks. This method should be similar to the `check_estimator` method that sklearn provides (https://scikit-learn.org/stable/modules/generated/sklearn.utils.estimator_checks.check_estimator.html).

### PR DESCRIPTION
**Pull request by ravisharma-09**
# DO NOT REMOVE THIS CHECKLIST

### Your checklist for this pull request
Your PR **won't** get reviewed if this checklist is not present in the PR.

Please complete the following checklist after creating the PR.

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side) against our **dev branch** (left side). Please do not request your **dev branch**.
- [x] Have you followed all the steps from our [Contributing Guide](https://github.com/pgmpy/pgmpy/blob/dev/Contributing.md)?
- [ ] Are all the GitHub Actions checks passing? If not, they will need to be fixed before review. You can reference logs for the failing check to identify the issue.

Did you use a Large language model (LLM) to assist you in making changes? If yes, please go through the following checklist too:
- [x] Please include a short paragraph (**not bullet points**) describing the changes you have made. This should at least include the problem your PR solves and the implemented solution. This doesn't need to be a polished description, but it **must** be written **manually** by you without any assistance from any AI tools.
- [x] Have you **manually** verified that the changes are doing exactly what is expected? Have you considered edge cases?
- [x] Has the LLM added try-except blocks? They will need to be removed; any error handling must be explicit.
- [x] If you used LLM for generating tests, they usually need to be compressed into a smaller set of tests. Please **manually** describe what scenarios the tests are testing for.

### Issue number(s) that this pull request fixes
- Fixes #2839

### List of changes to the codebase in this pull request
- Developed `check_causal_discovery` inside `pgmpy/utils/estimator_checks.py` to validate new causal discovery algorithms against the standard pgmpy interface.
- Registered the function in `pgmpy/utils/__init__.py`.
- Added a dedicated test suite in `pgmpy/tests/test_utils/test_estimator_checks.py`.

### Manual Description
This PR solves the problem of ensuring that newly implemented causal discovery algorithms adhere to the unified `pgmpy` interface mapped out in the causal discovery template. I implemented a new utility function, `check_causal_discovery`, modeled after `sklearn.utils.estimator_checks`. The function validates that a given estimator has the required `fit` and `score` methods, properly sets interior attributes like `causal_graph_` and `n_features_in_` after being fitted on a test dataset, and correctly handles errors. To ensure reliability, test cases were created using a set of dummy estimators mimicking complete and incomplete logic, alongside verifying the standard `PC` algorithm instance. I have also manually run these tests locally using `pytest`, and all checks successfully pass.

### CODE REVIEW

**Review by ankurankan — CHANGES_REQUESTED** [review]
1. The `estimator_checks.py` files should be in `test_causal_discovery` dir.
2. Why have you committed a text file with commit message?
3. What have you commited `test_dir/test_file`?

**Comment by ravisharma-09** [comment]
Thanks for the review and for assigning the issue!

Regarding the questions:

* The text file with the commit message was accidentally committed while testing locally during development. It was not intended to be part of the PR.
* Similarly, `test_dir/test_file` was a temporary file used during early testing of the interface checks and was mistakenly included in the commit.

These files are not related to the `check_causal_discovery` implementation. I’ve now removed the unintended files and cleaned up the branch so that the PR only contains the relevant changes for issue #2839.

Please let me know if anything else needs adjustment.

File: pgmpy/tests/test_causal_discovery/test_check_causal_discovery.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+    def test_valid_estimator(self):
+        est5 = DummyEstimator5()
+        self.assertTrue(check_causal_discovery(est5))
+        est5.score()
+
+    def test_real_estimator(self):
+        pc = PC()
+        # The test checks whether PC correctly adheres to the interface without throwing errors
+        self.assertTrue(check_causal_discovery(pc))
```
Comment (ankurankan) [inline pgmpy/tests/test_causal_discovery/test_check_causal_discovery.py:156]: This shouldn't be in this file.

File: pgmpy/estimators/causal_discovery/check_causal_discovery.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            f"The `fit` method of {name} failed when run on a dummy dataset with error: {e}"
+        )
+
+    # 3. Check for specific properties that must be populated after run
+    if not hasattr(estimator, "causal_graph_"):
+        raise AttributeError(
+            f"The 'fit' method of {name} must set the internal attribute 'causal_graph_'."
+        )
+
+    if not hasattr(estimator, "n_features_in_"):
```
Comment (ankurankan) [inline pgmpy/estimators/causal_discovery/check_causal_discovery.py:55]: Can you make it stricter by also validating that this is a positive integer?

File: pgmpy/estimators/causal_discovery/check_causal_discovery.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            f"The 'fit' method of {name} must set the internal attribute 'n_features_in_'."
+        )
+
+    if not hasattr(estimator, "feature_names_in_"):
+        raise AttributeError(
+            f"The 'fit' method of {name} must set the internal attribute 'feature_names_in_'."
+        )
+
+    # 4. We can optionally verify that adjacency_matrix_ is populated
+    if not hasattr(estimator, "adjacency_matrix_"):
```
Comment (ankurankan) [inline pgmpy/estimators/causal_discovery/check_causal_discovery.py:66]: Make it stricter by checking that this is a 2 dim matrix with number of columns and rows matching exactly the number of columns in the passed dataset.

File: pgmpy/estimators/causal_discovery/check_causal_discovery.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        raise AttributeError(
+            f"The 'fit' method of {name} must set the internal attribute 'causal_graph_'."
+        )
+
+    if not hasattr(estimator, "n_features_in_"):
+        raise AttributeError(
+            f"The 'fit' method of {name} must set the internal attribute 'n_features_in_'."
+        )
+
+    if not hasattr(estimator, "feature_names_in_"):
```
Comment (ankurankan) [inline pgmpy/estimators/causal_discovery/check_causal_discovery.py:60]: Also validate that the size of this is exactly same as `n_features_in_`

**Review by ankurankan — CHANGES_REQUESTED** [review]
I have added a few comments. Looks mostly good to me. Can you also add a new test in each of the causal discovery algorithms' test files using this new estimator check method?

**Comment by ankurankan** [comment]
Closing due to inactivity.
